### PR TITLE
feat(usage): Add usage collector plugin

### PR DESCRIPTION
**Pull request by skarekrow**
This will collect anonymous data that can be aggregated.

NAS-101420

Signed-off-by: Brandon Schneider <brandon@ixsystems.com>

### CODE REVIEW

File: src/middlewared/middlewared/plugins/usage.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+                if not d.get('path'):
+                    vdevs += 1
+                    type = d['type']
+                    disks += len(d['children'])
+                else:
+                    disks += 1
+                    used = pd['usedbysnapshots']['parsed'] + \
+                        pd['usedbydataset']['parsed'] + \
+                        pd['usedbychildren']['parsed'] + \
+                        pd['usedbyrefreservation']['parsed']
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/usage.py]: I think I was not clear here,

this is the same as just `pd['usedbydataset']['parsed']`

Thought we could perhaps also send other usedby* to usage stats.

**Comment by william-gr** [comment]
@themylogin would love to know your opinion on all of this too ; )

File: src/middlewared/middlewared/plugins/usage.py [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+import random
+import requests
+
+
+class UsageService(Service):
+    @private
+    async def start(self):
+        try:
+            gather = await self.gather()
+            requests.post(url='https://usage.freenas.org/submit', data=gather)
```
[the remark this answers is not in the document]
Comment (themylogin) [reply src/middlewared/middlewared/plugins/usage.py]: Also need a timeout here, and `headers={"Content-type": "application/json"}`

File: src/middlewared/middlewared/plugins/usage.py [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+import asyncio
+import random
+import requests
+
+
+class UsageService(Service):
+    @private
+    async def start(self):
+        try:
+            gather = await self.gather()
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/usage.py]: Probably,. if `system.general` -> `crash_reporting` is false, we need to skip gathering and sending data. @william-gr what do you think?

File: src/middlewared/middlewared/plugins/usage.py [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+            requests.post(url='https://usage.freenas.org/submit', data=gather)
+        except Exception:
+            # We still want to schedule the next call
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/usage.py]: Log the exception please

File: src/middlewared/middlewared/plugins/usage.py (hunk 3 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/usage.py]: Yeah, we need a way to opt-out for this but I am not sure reusing crash_reporting is our best option. Will need to think about it.

Comment (themylogin) [reply src/middlewared/middlewared/plugins/usage.py]: So we are going to ship 11.3 with this always enabled?

Comment (william-gr) [reply src/middlewared/middlewared/plugins/usage.py]: No, we are still long ways to 11.3. We will figure something out until then.

**Pull request merged**